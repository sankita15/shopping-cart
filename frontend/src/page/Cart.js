/* eslint-disable react/jsx-no-target-blank */
import React from 'react';
import { Label, Media, Table } from 'reactstrap';
import { FaMinusSquare, FaPlusSquare } from 'react-icons/fa';
import PropTypes from 'prop-types';
import { getCartProducts, getCartResponse, getTotalPrice, isCartEmpty } from './CartAvailability';


export default class Cart extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            cartDetails: {},
        };

        this.getProduct = this.getProduct.bind(this);
    }

    componentDidMount() {
        const { user } = this.props;

        getCartResponse(user)
            .then(cartDetails => this.setState({ cartDetails }))
            .catch(status => console.warn(status));
    }

    getProductLink = productId => `/products/${productId}`;

    getProduct({ id, imageUrl, productName, price }, productQuantity) {
        return (
            <tr key={id}>
                <td>
                    <a target="_blank" href={this.getProductLink(id)}>
                        <Media className="cart-image" src={imageUrl} />
                    </a>
                </td>
                <td>
                    <a target="_blank" className="cart-product-name" href={this.getProductLink(id)}>
                        {productName}
                    </a>
                </td>
                <td>
                    <Label>
                    Rs.
                        {' '}
                        {price}
                    </Label>
                </td>
                <td>
                    <FaMinusSquare className="cart-minus" onClick={() => this.updateProductInCart(id, 'DELETE')} />
                    <Label className="cart-quantity-label">
                        {productQuantity}
                    </Label>
                    <FaPlusSquare className="cart-plus" onClick={() => this.updateProductInCart(id, 'POST')} />
                </td>
            </tr>
        );
    }

    render() {
        const { cartDetails } = this.state;

        if (!isCartEmpty(cartDetails)) {
            return <Label>Your Cart is empty.Please add item to your cart</Label>;
        }

        return (
            <div>
                <Label>Your Shopping cart</Label>
                <Table>
                    <thead>
                        <tr>
                            <th>Product</th>
                            <th>Name</th>
                            <th>Price</th>
                            <th>Quantity</th>
                        </tr>
                    </thead>
                    <tbody>{getCartProducts(cartDetails, this.getProduct)}</tbody>
                </Table>
                <div className="table-row">
                    <Label className="table-final-row">
                        {`SubTotal: ${getTotalPrice(cartDetails.totalPrice)}`}
                    </Label>
                </div>
            </div>
        );
    }

    updateProductInCart(productId, methodName) {
        const { cartDetails: { id: cartId } } = this.state;

        fetch(`/api/carts/${cartId}/product/${productId}`, {
            credentials: 'include',
            method: methodName,
            headers: {
                'Content-Type': 'application/json',
            },
        }).then(res => (res.ok ? res.json() : Promise.reject(res.status)))
            .then(cartDetails => this.setState({ cartDetails }))
            .catch(status => console.warn(status));
    }
}

Cart.propTypes = {
    user: PropTypes.string.isRequired,
};
