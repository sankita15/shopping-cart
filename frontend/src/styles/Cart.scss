.cart-image{
  height: 81px;
  margin-right: -106px;
  cursor: pointer;
}
.cart-product-name{
  cursor: pointer;
  color: black;
}
.cart-plus{
  cursor: pointer;
}
.cart-minus{
  cursor: pointer;
}
.cart-quantity-label{
  margin-left: 8px;
  margin-right: 8px;
}
.table-row{
  background-color: lightgrey;
  height: 1px;
}
.table-final-row{
  font-weight: bold;
  margin-left: 1071px;
  margin-top: 15px;
}
.table td{
  vertical-align: middle;
}
