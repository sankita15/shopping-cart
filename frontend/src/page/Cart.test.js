import React from 'react';
import { shallow } from 'enzyme';
import { Table, Label } from 'reactstrap';
import { FaPlusSquare, FaMinusSquare } from 'react-icons/fa';
import Cart from './Cart';

describe('carts page', () => {
    const defaultProps = {
        user: 'alice',
    };

    const EMPTY_CART = [
        {
            id: '5c6bf8a306057eeb12dfb7c4',
            status: 'pending',
            username: 'alice',
            products: {},
            productQuantities: {},
            lastModified: 1551162037567,
            orderDate: 1550482348428,
            totalPrice: 547,
        },
    ];

    const CART = [
        {
            id: '5c6bf8a306057eeb12dfb7c4',
            status: 'pending',
            username: 'alice',
            products: {
                '59d89875479ded2a718bac13': {
                    'id': '59d89875479ded2a718bac13',
                    'productName': 'colgatewhite',
                    'productCode': 'GE-1206',
                    'releaseDate': 1483539355570,
                    'lastModified': 1507367029712,
                    'description': 'dentifrice',
                    'price': 300,
                    'starRating': 4.5,
                    'imageUrl': 'http://colgate.com/2/',
                    'stock': 1000,
                },
                '59d89875479ded2a718bac14': {
                    'id': '59d89875479ded2a718bac14',
                    'productName': 'gilettemousse',
                    'productCode': 'GE-1206',
                    'releaseDate': 1486224955570,
                    'lastModified': 1507367029713,
                    'description': 'moussearaser',
                    'price': 200,
                    'starRating': 4.2,
                    'imageUrl': 'http://gilette.com/2/',
                    'stock': 1000,
                },
            },
            productQuantities: {
                '59d89875479ded2a718bac13': 2,
                '59d89875479ded2a718bac14': 2,
            },
            lastModified: 1551162037567,
            orderDate: 1550482348428,
            totalPrice: 1000,
        },
    ];

    const SECOND_CART = {
        id: '5c6bf8a306057eeb12dfb7c4',
        status: 'pending',
        username: 'alice',
        products: {
            '59d89875479ded2a718bac13': {
                'id': '59d89875479ded2a718bac13',
                'productName': 'colgatewhite',
                'productCode': 'GE-1206',
                'releaseDate': 1483539355570,
                'lastModified': 1507367029712,
                'description': 'dentifrice',
                'price': 3,
                'starRating': 4.5,
                'imageUrl': 'http://colgate.com/2/',
                'stock': 1000,
            },
            '59d89875479ded2a718bac14': {
                'id': '59d89875479ded2a718bac14',
                'productName': 'gilettemousse',
                'productCode': 'GE-1206',
                'releaseDate': 1486224955570,
                'lastModified': 1507367029713,
                'description': 'moussearaser',
                'price': 2,
                'starRating': 4.2,
                'imageUrl': 'http://gilette.com/2/',
                'stock': 1000,
            },
        },
        productQuantities: {
            '59d89875479ded2a718bac13': 3,
            '59d89875479ded2a718bac14': 2,
        },
        lastModified: 1551162037567,
        orderDate: 1550482348428,
        totalPrice: 1300,
    };

    const ORDERED_CART = [{
        id: '5c6bf8a306057eeb12dfb7c4',
        status: 'ordered',
        username: 'alice',
        products: {
            '59d89875479ded2a718bac13': {
                'id': '59d89875479ded2a718bac13',
                'productName': 'colgatewhite',
                'productCode': 'GE-1206',
                'releaseDate': 1483539355570,
                'lastModified': 1507367029712,
                'description': 'dentifrice',
                'price': 300,
                'starRating': 4.5,
                'imageUrl': 'http://colgate.com/2/',
                'stock': 1000,
            },
        },
        productQuantities: {
            '59d89875479ded2a718bac13': 2,
        },
        lastModified: 1551162037567,
        orderDate: 1550482348428,
        totalPrice: 1000,
    },
    {
        id: '5c7ab5bf06057e568d0f76c4',
        status: 'ordered',
        username: 'alice',
        products: {
            '5c7ab0e006057e50bbc8d898': {
                'id': '5c7ab0e006057e50bbc8d898',
                'productName': 'Garden Cart',
                'productCode': 'GDN-0023',
                'releaseDate': 1489805155570,
                'lastModified': 1551544544142,
                'description': '15 gallon capacity rolling garden cart',
                'price': 3295,
                'starRating': 4.2,
                'imageUrl': 'http://openclipart.org/image/300px/svg_to_png/58471/garden_cart.png',
                'stock': 14,
            },
        },
        productQuantities: {
            '5c7ab0e006057e50bbc8d898': 1,
        },
        lastModified: 1551545797603,
        orderDate: 1551545797603,
        totalPrice: 3295,
    }];

    const THIRD_CART = {
        id: '5c6bf8a306057eeb12dfb7c4',
        status: 'pending',
        products: {
            '59d89875479ded2a718bac13': {
                'id': '59d89875479ded2a718bac13',
                'productName': 'colgatewhite',
                'productCode': 'GE-1206',
                'releaseDate': 1483539355570,
                'lastModified': 1507367029712,
                'description': 'dentifrice',
                'price': 3,
                'starRating': 4.5,
                'imageUrl': 'http://colgate.com/2/',
                'stock': 1000,
            },
            '59d89875479ded2a718bac14': {
                'id': '59d89875479ded2a718bac14',
                'productName': 'gilettemousse',
                'productCode': 'GE-1206',
                'releaseDate': 1486224955570,
                'lastModified': 1507367029713,
                'description': 'moussearaser',
                'price': 2,
                'starRating': 4.2,
                'imageUrl': 'http://gilette.com/2/',
                'stock': 1000,
            },
        },
        productQuantities: {
            '59d89875479ded2a718bac13': 1,
            '59d89875479ded2a718bac14': 2,
        },
        lastModified: 1551162037567,
        orderDate: 1550482348428,
        totalPrice: 700,
    };

    let component;

    const createWrapper = () => shallow(<Cart {...defaultProps} />);

    beforeEach(async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(CART), ok: true });

        component = createWrapper();

        // eslint-disable-next-line no-undef
        await flushPromises();
    });


    afterEach(() => {
        jest.clearAllMocks();
        fetch.resetMocks();
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should render carts page when CART is not empty', () => {
        const table = component.find(Table);

        expect(table.length).toBe(1);
    });

    it('should call fetch for rendering carts data', () => {
        expect(fetch).toHaveBeenCalledWith(`/api/carts/user/${defaultProps.user}`, {
            credentials: 'include',
            headers: {
                'Content-Type': 'application/json',
            },
        });
    });

    it('should render message when CART is not created', async () => {
        fetch.resetMocks();

        fetch.mockResolvedValueOnce({ json: () => Promise.resolve([{}]), ok: true });

        const wrapper = createWrapper();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const message = wrapper.find(Label);

        expect(message.length).toBe(1);
    });

    it('should render message when CART is created but empty', async () => {
        fetch.resetMocks();

        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(EMPTY_CART), ok: true });

        const wrapper = createWrapper();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const message = wrapper.find(Label);

        expect(message.length).toBe(1);
    });

    it('should render Cart Empty message when there is no pending cart available', async () => {
        fetch.resetMocks();

        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(ORDERED_CART), ok: true });

        const wrapper2 = createWrapper();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const message = wrapper2.find(Label);

        expect(message.length).toBe(1);
    });

    it('should redirect to product page on click of product name', () => {
        const table = component.find(Table);
        const productName = table.find('a').at(1).prop('href');
        const productId = Object.keys(CART[0].products)[0];

        expect(productName).toBe(`/products/${productId}`);
    });

    it('should redirect to product page on click of product image', () => {
        const table = component.find(Table);
        const productImage = table.find('a').at(0).prop('href');
        const productId = Object.keys(CART[0].products)[0];

        expect(productImage).toBe(`/products/${productId}`);
    });

    it('should add product quantity on click of add icon', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(SECOND_CART), ok: true });

        const table = component.find(Table);
        const addIcon = table.find(FaPlusSquare).at(0);
        const previousQuantity = table.find(Label).at(1);

        expect(previousQuantity.prop('children')).toEqual(2);

        addIcon.simulate('click');

        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        expect(fetch).toHaveBeenCalledWith('/api/carts/5c6bf8a306057eeb12dfb7c4/product/59d89875479ded2a718bac13', {
            credentials: 'include',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
        });

        // eslint-disable-next-line no-undef
        await flushPromises();

        const quantity = component.find(Table).find(Label).at(1);

        expect(quantity.prop('children')).toEqual(3);
    });

    it('should reduce the count of product on click on minus icon', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(THIRD_CART), ok: true });
        const table = component.find(Table);
        const subtractIcon = component.find(FaMinusSquare).at(0);
        const previousQuantity = table.find(Label).at(1);

        expect(previousQuantity.prop('children')).toEqual(2);

        subtractIcon.simulate('click');

        component.update();

        expect(fetch).toHaveBeenCalledWith('/api/carts/5c6bf8a306057eeb12dfb7c4/product/59d89875479ded2a718bac13', {
            credentials: 'include',
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json',
            },
        });

        // eslint-disable-next-line no-undef
        await flushPromises();

        const quantity = component.find(Table).find(Label).at(1);

        expect(quantity.prop('children')).toEqual(1);
    });

    it('should render total price after adding new item', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(SECOND_CART), ok: true });
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(THIRD_CART), ok: true });

        const table = component.find(Table);
        const addIcon = table.find(FaPlusSquare).at(0);
        const subtractIcon = table.find(FaMinusSquare).at(0);

        const initialPrice = component.find(Label).at(5);

        expect(initialPrice.props().children).toEqual(`SubTotal: ${CART[0].totalPrice}`);

        addIcon.simulate('click');
        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const previousTotalPrice = component.find(Label).at(5);

        expect(previousTotalPrice.props().children).toEqual(`SubTotal: ${SECOND_CART.totalPrice}`);

        subtractIcon.simulate('click');
        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        const totalPrice = component.find(Label).at(5);

        expect(totalPrice.prop('children')).toEqual(`SubTotal: ${THIRD_CART.totalPrice}`);
    });
});
