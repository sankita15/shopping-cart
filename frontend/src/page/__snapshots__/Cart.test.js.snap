// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carts page should match snapshot 1`] = `
<div>
  <Label
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    Your Shopping cart
  </Label>
  <Table
    responsiveTag="div"
    tag="table"
  >
    <thead>
      <tr>
        <th>
          Product
        </th>
        <th>
          Name
        </th>
        <th>
          Price
        </th>
        <th>
          Quantity
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="59d89875479ded2a718bac13"
      >
        <td>
          <a
            href="/products/59d89875479ded2a718bac13"
            target="_blank"
          >
            <Media
              className="cart-image"
              src="http://colgate.com/2/"
            />
          </a>
        </td>
        <td>
          <a
            className="cart-product-name"
            href="/products/59d89875479ded2a718bac13"
            target="_blank"
          >
            colgatewhite
          </a>
        </td>
        <td>
          <Label
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Rs.
             
            300
          </Label>
        </td>
        <td>
          <FaMinusSquare
            className="cart-minus"
            onClick={[Function]}
          />
          <Label
            className="cart-quantity-label"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            2
          </Label>
          <FaPlusSquare
            className="cart-plus"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        key="59d89875479ded2a718bac14"
      >
        <td>
          <a
            href="/products/59d89875479ded2a718bac14"
            target="_blank"
          >
            <Media
              className="cart-image"
              src="http://gilette.com/2/"
            />
          </a>
        </td>
        <td>
          <a
            className="cart-product-name"
            href="/products/59d89875479ded2a718bac14"
            target="_blank"
          >
            gilettemousse
          </a>
        </td>
        <td>
          <Label
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Rs.
             
            200
          </Label>
        </td>
        <td>
          <FaMinusSquare
            className="cart-minus"
            onClick={[Function]}
          />
          <Label
            className="cart-quantity-label"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            2
          </Label>
          <FaPlusSquare
            className="cart-plus"
            onClick={[Function]}
          />
        </td>
      </tr>
    </tbody>
  </Table>
  <div
    className="table-row"
  >
    <Label
      className="table-final-row"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      SubTotal: 1000
    </Label>
  </div>
</div>
`;
